core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -143,6 +143,7 @@ private[log] class LogCleanerManager(val logDirs: Array[File], val logs: Pool[To
           inProgress.put(topicAndPartition, LogCleaningPaused)
         case Some(state) =>
           state match {
+            case LogCleaningPaused => {}
```

Comment:
@junrao  We've also hit this error while reassigning partitions. As you hinted, just before the `java.lang.IllegalStateException` we see this:
```
[2017-03-23 13:24:06,081] WARN [ReplicaFetcherThread-0-1], Replica 2 for partition [testpfx2hn-hyperion-native-4,0] reset its fetch offset from 0 to current leader 1's latest offset -1 (kafka.server.Replic
aFetcherThread)
[2017-03-23 13:24:06,081] INFO Truncating log testpfx2hn-hyperion-native-4-0 to offset -1. (kafka.log.Log)
[2017-03-23 13:24:06,082] ERROR [ReplicaFetcherThread-0-1], Error getting offset for partition [testpfx2hn-hyperion-native-4,0] to broker 1 (kafka.server.ReplicaFetcherThread)
java.lang.IllegalArgumentException: Cannot truncate to a negative offset (-1).
        at kafka.log.Log.truncateTo(Log.scala:731)
        at kafka.log.LogManager$$anonfun$truncateTo$2.apply(LogManager.scala:288)
        at kafka.log.LogManager$$anonfun$truncateTo$2.apply(LogManager.scala:280)
        at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:778)
        at scala.collection.immutable.Map$Map1.foreach(Map.scala:116)
        at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:777)
        at kafka.log.LogManager.truncateTo(LogManager.scala:280)
        at kafka.server.ReplicaFetcherThread.handleOffsetOutOfRange(ReplicaFetcherThread.scala:186)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2$$anonfun$apply$mcV$sp$1$$anonfun$apply$2.apply(AbstractFetcherThread.scala:157)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2$$anonfun$apply$mcV$sp$1$$anonfun$apply$2.apply(AbstractFetcherThread.scala:127)
        at scala.Option.foreach(Option.scala:257)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2$$anonfun$apply$mcV$sp$1.apply(AbstractFetcherThread.scala:127)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2$$anonfun$apply$mcV$sp$1.apply(AbstractFetcherThread.scala:125)
        at scala.collection.mutable.HashMap$$anonfun$foreach$1.apply(HashMap.scala:99)
        at scala.collection.mutable.HashMap$$anonfun$foreach$1.apply(HashMap.scala:99)
        at scala.collection.mutable.HashTable$class.foreachEntry(HashTable.scala:230)
        at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:40)
        at scala.collection.mutable.HashMap.foreach(HashMap.scala:99)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2.apply$mcV$sp(AbstractFetcherThread.scala:125)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2.apply(AbstractFetcherThread.scala:125)
        at kafka.server.AbstractFetcherThread$$anonfun$processFetchRequest$2.apply(AbstractFetcherThread.scala:125)
        at kafka.utils.CoreUtils$.inLock(CoreUtils.scala:231)
        at kafka.server.AbstractFetcherThread.processFetchRequest(AbstractFetcherThread.scala:123)
        at kafka.server.AbstractFetcherThread.doWork(AbstractFetcherThread.scala:98)
        at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:63)
```

This is probably the root cause.
Just for the record, we've tried the patch and while I agree it doesn't fix the root cause, it worked for us.